datafusion/physical-expr/src/expressions/datetime.rs

```diff
@@ -86,76 +89,114 @@ impl PhysicalExpr for DateIntervalExpr {
-                ScalarValue::Date64(Some(date)) => match &self.op {
-                    Operator::Plus => Ok(ColumnarValue::Scalar(ScalarValue::Date64(
-                        Some(date + interval as i64),
-                    ))),
-                    Operator::Minus => Ok(ColumnarValue::Scalar(ScalarValue::Date64(
-                        Some(date - interval as i64),
-                    ))),
-                    _ => {
-                        // this should be unreachable because we check the operators in `try_new`
-                        Err(DataFusionError::Execution(
-                            "Invalid operator for DateIntervalExpr".to_string(),
-                        ))
-                    }
-                },
-                _ => {
-                    // this should be unreachable because we check the types in `try_new`
-                    Err(DataFusionError::Execution(
-                        "Invalid lhs type for DateIntervalExpr".to_string(),
-                    ))
-                }
-            },
+        // Unwrap interval to add
+        let scalar = match &intervals {
+            ColumnarValue::Scalar(interval) => interval,
             _ => Err(DataFusionError::Execution(
                 "Columnar execution is not yet supported for DateIntervalExpr"
                     .to_string(),
-            )),
-        }
+            ))?,
+        };
+
+        // Negate for subtraction
+        let interval = match &scalar {
+            ScalarValue::IntervalDayTime(Some(interval)) => *interval,
+            ScalarValue::IntervalYearMonth(Some(interval)) => *interval as i64,
+            ScalarValue::IntervalMonthDayNano(Some(_interval)) => {
+                Err(DataFusionError::Execution(
+                    "DateIntervalExpr does not support IntervalMonthDayNano".to_string(),
+                ))?
+            }
+            other => Err(DataFusionError::Execution(format!(
+                "DateIntervalExpr does not support non-interval type {:?}",
+                other
+            )))?,
+        };
+        let interval = match &self.op {
+            Operator::Plus => interval,
+            Operator::Minus => -interval,
+            _ => {
+                // this should be unreachable because we check the operators in `try_new`
+                Err(DataFusionError::Execution(
+                    "Invalid operator for DateIntervalExpr".to_string(),
+                ))?
+            }
+        };
+
+        // Add interval
+        let posterior = match scalar {
+            ScalarValue::IntervalDayTime(Some(_)) => {
+                prior.add(chrono::Duration::days(interval))
+            }
+            ScalarValue::IntervalYearMonth(Some(_)) => {
+                let target = add_months(prior, interval);
+                let target_plus = add_months(target, 1);
+                let last_day = target_plus.sub(chrono::Duration::days(1));
+                let day = min(prior.day(), last_day.day());
```

Comment:
I don't really understand this logic -- I would have expected that code would split the interval into the 3 three fields (month, day, nano) and then use the various Chrono operations to do the timestamp arithmetic

https://github.com/apache/arrow/blob/master/format/Schema.fbs#L354-L375

Reply:
When the type is `IntervalDayTime`, I think I copied the behavior of the existing code, which was to grab a 32-bit number, which appears to represent the number of days based on this snippet https://github.com/apache/arrow-datafusion/blob/7617d78809d4ff5bde31142e0744c70024e40635/datafusion/physical-expr/src/expressions/datetime.rs#L122 (I assume there were tests around this? and if so they still pass).

For `IntervalYearMonth` I guessed and was correct based on the documentation you linked (ty!):
```
YEAR_MONTH - Indicates the number of elapsed whole months
```

I wasn't clear on what `IntervalMonthDayNano` meant, but now I do thanks to your link! This PR leaves it unimplemented (no change in behavior). It isn't critical to passing the TPC-H benchmarks, so it wasn't a priority for me at this moment.

As for the actual business logic of `IntervalYearMonth`, it is always stored as a number of months (year=12), and it turns out that adding (and subtracting) months is not very straight forward, and not implemented in chrono https://stackoverflow.com/questions/64081289/how-do-i-add-a-month-to-a-chrono-naivedate . When I looked into other implementations, they appeared to have issues (i.e. not supporting negative arguments), so I wrote this.

Unfortunately, one can't just add a fixed time unit (months have different numbers of days), and one has to know what is being added to in order to add properly (some Februaries have 29 days, others only 28). Observing postgres behavior, days of the month get clamped `2022-01-31` + 1 month = `2022-02-28`, so this logic is attempting to reproduce that behavior.

Reply:
Reflecting, there could be a valid argument to PR chrono :thinking:

Reply:
Thank you @avantgardnerio

Reply:
I implemented `IntervalDayTime` and `IntervalMonthDayNano`, but could only figure out how to test `IntervalDayTime`, since I need a `timestamp` to test < 1 day resolutions, and it didn't seem like the timestamp path was plumbed or was plumbing it in scope for this PR.